### PR TITLE
MaskFormer, Mask2Former - replace einsum for tracing

### LINKED ISSUE #25261: Mask2Former broadcasting issue when running inference on model traced with GPU device
### System Info

```
- System information: x86_64 GNU/Linux
- Ubuntu version: 18.04
- Python version: 3.8.12
- CUDA version: 11.1
- PyTorch version: 2.0.1
- transformers version: 4.31.0
```

### Who can help?

@amyeroberts
@sgugger 
@muellerzr

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
import torch
from transformers import Mask2FormerForUniversalSegmentation

device = torch.device("cuda")
model = Mask2FormerForUniversalSegmentation.from_pretrained(
                "facebook/mask2former-swin-tiny-coco-instance",
                torchscript=True
                ).eval().to(device)
dummy_input = torch.randn((1,3,640,640)).to(device)
traced_model = torch.jit.trace(model, dummy_input)

with torch.no_grad():
    out = traced_model(torch.randn((2,3,640,640)).to(device))
    out = traced_model(torch.randn((2,3,640,640)).to(device))
```

The above code generates the following error when calling the **second** forward of `traced_model` (last line):
```
Traceback (most recent call last):
  File "mask2former_trace.py", line 14, in <module>
    out = traced_model(torch.randn((2,3,640,640)).to(device))
  File "~/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
~/python3.8/site-packages/torch/functional.py(378): einsum
~/python3.8/site-packages/transformers/models/mask2former/modeling_mask2former.py(2015): forward
~/python3.8/site-packages/torch/nn/modules/module.py(1488): _slow_forward
~/python3.8/site-packages/torch/nn/modules/module.py(1501): _call_impl
~/python3.8/site-packages/transformers/models/mask2former/modeling_mask2former.py(1852): forward
~/python3.8/site-packages/torch/nn/modules/module.py(1488): _slow_forward
~/python3.8/site-packages/torch/nn/modules/module.py(1501): _call_impl
~/python3.8/site-packages/transformers/models/mask2former/modeling_mask2former.py(2080): forward
~/python3.8/site-packages/torch/nn/modules/module.py(1488): _slow_forward
~/python3.8/site-packages/torch/nn/modules/module.py(1501): _call_impl
~/python3.8/site-packages/transformers/models/mask2former/modeling_mask2former.py(2271): forward
~/python3.8/site-packages/torch/nn/modules/module.py(1488): _slow_forward
~/python3.8/site-packages/torch/nn/modules/module.py(1501): _call_impl
~/python3.8/site-packages/transformers/models/mask2former/modeling_mask2former.py(2496): forward
~/python3.8/site-packages/torch/nn/modules/module.py(1488): _slow_forward
~/python3.8/site-packages/torch/nn/modules/module.py(1501): _call_impl
~/python3.8/site-packages/torch/jit/_trace.py(1056): trace_module
~/python3.8/site-packages/torch/jit/_trace.py(794): trace
mask2former_trace.py(10): <module>
RuntimeError: einsum(): subscript b has size 2 for operand 1 which does not broadcast with previously seen size 400
```

If I trace the model with batch size 2, i.e. `dummy_input = torch.randn((2,3,640,640)).to(device)`, the same error arises at the **first** forward call of `traced_model`

The issue seems to be [here](https://github.com/huggingface/transformers/blob/e42587f596181396e1c4b63660abf0c736b10dae/src/transformers/models/mask2former/modeling_mask2former.py#L2015)

### Expected behavior

When tracing on CPU, i.e. in the code above:
```
device = torch.device("cpu")
```
everything works fine. I would expect similar behaviour when tracing on GPU device.

**Additional notes**: 
I already tried tracing the model on CPU device, then moving `traced_model` (as well as the input tensors) to GPU, and running inference, but I got the following error:
```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```
I know this is a known issue:
https://github.com/huggingface/transformers/issues/5664
https://github.com/huggingface/transformers/issues/22038
so I guess there should be some tensors in Mask2Former created at forward time with the same device as the input, and torchscript does not change that device when running on GPU. 
This is the reason why I need to trace the model on GPU.

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Maskformer cannot currently be traced because of einsum operations. This PR replaces the einsum operations with standard matmuls.  

With this PR, the following now runs: 

```python
import torch
from transformers import Mask2FormerForUniversalSegmentation

device = torch.device("cuda")
model = Mask2FormerForUniversalSegmentation.from_pretrained(
    "facebook/mask2former-swin-tiny-coco-instance",
    torchscript=True
).eval().to(device)
dummy_input = torch.randn((1,3,640,640)).to(device)
traced_model = torch.jit.trace(model, dummy_input)

with torch.no_grad():
    out = traced_model(torch.randn((2,3,640,640)).to(device))
    out = traced_model(torch.randn((2,3,640,640)).to(device))
```

Partially resolves #25261  - enables tracing but does not resolve the issue of different results between traced and non-traced model on GPU

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: src/transformers/models/maskformer/modeling_maskformer.py [hunk 1]
```diff
@@ -1727,7 +1733,12 @@ def get_logits(self, outputs: MaskFormerModelOutput) -> Tuple[Tensor, Tensor, Di
             # get the masks
             mask_embeddings = self.mask_embedder(transformer_decoder_hidden_states)
             # sum up over the channels
-            masks_queries_logits = torch.einsum("bqc,   bchw -> bqhw", mask_embeddings, pixel_embeddings)
+            # (batch_size, num_queries, num_channels, 1, 1)
+            mask_embeddings = mask_embeddings.unsqueeze(-1).unsqueeze(-1)
+            # (batch_size, 1, num_channels, height, width)
+            pixel_embeddings = pixel_embeddings.unsqueeze(1)
+            # (batch_size, num_queries, num_channels, height, width)
```
Comment (amyeroberts) [inline src/transformers/models/maskformer/modeling_maskformer.py:1740]:
```suggestion
            # (batch_size, num_queries, height, width)
```

**Review by ydshieh — APPROVED** [review]
Thank you for working on this.

**Comment by ydshieh** [comment]
Hi @amyeroberts 

So the issue only happens at the 2nd `out = traced_model(torch.randn((2,3,640,640)).to(device))`. Before that (last) line, everything is fine. I am a dumb on this topic, but is it from the enisum? That's really weird (to me).

Still good for me, I am just going to run the torchscript tests and let you know.

**Comment by amyeroberts** [comment]
@ydshieh Yes, it's bizarre, although it does work on torch nightly. If you run the script above, it complains about the batch dimension in the einsum operation, and removing einsum resolved it, so I'm pretty sure that's the cause. I just don't know why 🤷‍♀️ 

For setting `fx_compatible = True`, I didn't add initially because the tests would fail on the torchscipt runs even if I added it to one of the [compatible models](https://github.com/huggingface/transformers/blob/a6e6b1c622d8d08e2510a82cb6266d7b654f1cbf/src/transformers/utils/fx.py#L113). 

Essentially, the model can be traced using `torch.jit.trace` if you accept many warnings, but fails when we use the `HFTracer` to trace. The tracer warnings and failings when using `HFTracer` are the same: Mask2Former tries to iterate over tensors which breaks things e.g. [here](https://github.com/huggingface/transformers/blob/a6e6b1c622d8d08e2510a82cb6266d7b654f1cbf/src/transformers/models/mask2former/modeling_mask2former.py#L815). Resolving these is a bigger piece of work - I'm happy to add them to this PR or in a follow up. It's possible these are related to the [difference between traced and non-traced models on GPU](https://github.com/huggingface/transformers/issues/25261#issuecomment-1663526119).

Were you able to run the tests successfully? There could be something funny in my env too.

**Comment by ydshieh** [comment]
Hi @amyeroberts Thanks for the info.

Forget about my comment about `fx_compatible = True`. The passed CI I mentioned is just the `test_torchscript_` tests (not the `test_torch_fx` - I didn't try as I saw it's not the raw `torch.jit.trace`)

**Pull request merged**